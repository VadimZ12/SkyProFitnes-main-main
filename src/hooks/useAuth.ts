import { useState, useEffect } from 'react';
import { User } from '../types/interfaces';
import { auth, database } from '../config/firebase';
import { onAuthStateChanged, signOut, updateProfile as updateFirebaseProfile, User as FirebaseUser } from 'firebase/auth';
import { ref, get, set } from "firebase/database";
import { registerUser, loginUser, resetPassword, changePassword } from '../api/auth';

const INACTIVITY_TIMEOUT = 3 * 60 * 1000; // 3 минуты в миллисекундах

export const useAuth = () => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);

  const updateLastActivity = () => {
    localStorage.setItem('lastActivity', Date.now().toString());
  };

  const checkInactivity = () => {
    const lastActivity = localStorage.getItem('lastActivity');
    if (lastActivity) {
      const inactiveTime = Date.now() - parseInt(lastActivity);
      if (inactiveTime > INACTIVITY_TIMEOUT) {
        logout();
      }
    }
  };

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (firebaseUser: FirebaseUser | null) => {
      try {
        if (firebaseUser) {
          checkInactivity();
          updateLastActivity();
          const userRef = ref(database, `users/${firebaseUser.uid}`);
          const snapshot = await get(userRef);
          if (snapshot.exists()) {
            const userData = snapshot.val();
            setUser({
              uid: firebaseUser.uid,
              email: firebaseUser.email!,
              displayName: userData.displayName || firebaseUser.displayName,
              customDisplayName: userData.customDisplayName,
            });
          } else {
            setUser({
              uid: firebaseUser.uid,
              email: firebaseUser.email!,
              displayName: firebaseUser.displayName || undefined,
            });
          }
        } else {
          setUser(null);
        }
      } catch (error) {
        console.error('Error in auth state change:', error);
        setUser(null);
      } finally {
        setLoading(false);
      }
    });

    window.addEventListener('mousemove', updateLastActivity);
    window.addEventListener('keypress', updateLastActivity);

    return () => {
      unsubscribe();
      window.removeEventListener('mousemove', updateLastActivity);
      window.removeEventListener('keypress', updateLastActivity);
    };
  }, []);

  const register = async (email: string, password: string) => {
    try {
      const user = await registerUser(email, password);
      updateLastActivity();
      setUser({
        uid: user.uid,
        email: user.email!,
        displayName: user.displayName || undefined,
      });
    } catch (error) {
      console.error('Error in register:', error);
      throw error;
    }
  };

  const login = async (email: string, password: string) => {
    try {
      const user = await loginUser(email, password);
      updateLastActivity();
      setUser({
        uid: user.uid,
        email: user.email!,
        displayName: user.displayName || undefined,
      });
    } catch (error) {
      console.error('Error in login:', error);
      throw error;
    }
  };

  const logout = async () => {
    try {
      await signOut(auth);
      localStorage.removeItem('lastActivity');
      setUser(null);
    } catch (error) {
      console.error('Error in logout:', error);
      throw error;
    }
  };

  const resetUserPassword = async (email: string) => {
    try {
      await resetPassword(email);
    } catch (error) {
      console.error('Error in resetPassword:', error);
      throw error;
    }
  };

  const changeUserPassword = async (newPassword: string) => {
    try {
      await changePassword(newPassword);
    } catch (error) {
      console.error('Error in changePassword:', error);
      throw error;
    }
  };

  const updateUser = async (updatedUser: Partial<User>) => {
    if (auth.currentUser) {
      try {
        const userRef = ref(database, `users/${auth.currentUser.uid}`);
        await updateFirebaseProfile(auth.currentUser, {
          displayName: updatedUser.displayName,
        });
        await set(userRef, updatedUser);
        setUser(currentUser => {
          if (currentUser) {
            return { ...currentUser, ...updatedUser };
          }
          return currentUser;
        });
      } catch (error) {
        console.error('Error updating user profile:', error);
        throw error;
      }
    }
  };

  return {
    user,
    loading,
    register,
    login,
    logout,
    resetUserPassword,
    changeUserPassword,
    updateUser,
  };
};