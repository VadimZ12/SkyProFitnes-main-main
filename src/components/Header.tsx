import React, { useState, useRef } from 'react';
import Button from '../components/Button';
import headerLogo from '/logo.png';
import { useAuth } from '../hooks/useAuth';
import { Link } from 'react-router-dom';
import UserModal from './Modals/UserModal';
interface HeaderProps {
  onLoginClick: () => void;
}

function Header({ onLoginClick }: HeaderProps) {
  const { user } = useAuth();
  const [isUserModalOpen, setUserModalOpen] = useState(false);
  const userRef = useRef<HTMLDivElement>(null);

  const handleUserClick = async (e: React.MouseEvent) => {
    if (e && e.stopPropagation) {
    e.stopPropagation(); // Предотвращаем распространение события
    setUserModalOpen(!isUserModalOpen);
    } // Переключаем состояние
  };

  return (
    <header className="flex justify-between mt-[50px] items-center">
      <Link to="/" className='flex flex-col gap-[15px]'>
        <img src={headerLogo} className="w-[220px] h-[35px]" alt="logo" loading="lazy"/>
        
        <p className="font-roboto text-[18px] font-normal leading-[19.8px] text-left text-gray mobile:hidden">
          Онлайн-тренировки для занятий дома
        </p>
      </Link>
      <div className='relative'>
        {user ? (
          <div
            ref={userRef}
            onClick={handleUserClick}
            className="flex items-center cursor-pointer gap-5"
          >
            <img
              src={user.photoURL || '/images/profile_no-img.png'}
              className="w-[41.67px] h-[41.67px] p-1 rounded-full bg-[#D9D9D9] object-cover"
              alt="user avatar"
            />
            <p className="block mobile:hidden font-roboto text-[24px] font-normal leading-[110%] text-black">
              {user.displayName || user.email?.split('@')[0]}
            </p>
            <svg className="w-[13px] h-[13px]" viewBox="0 0 14 7" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path d="M2 1L7 6L13 1" stroke="black" strokeWidth="2" strokeLinecap="round" strokeLinejoin="round" />
            </svg>
          </div>
        ) : (
          <Button onClick={onLoginClick} variant="primary" className="w-[103px] h-[52px]">Войти</Button>
        )}
        {user && isUserModalOpen && (
          <UserModal
            isOpen={isUserModalOpen}
            onClose={() => setUserModalOpen(false)}
            userRef={userRef}
          />
        )}
      </div>
    </header>
  );
}

export default Header;