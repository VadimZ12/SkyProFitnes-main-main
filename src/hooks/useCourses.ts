import { useState, useEffect, useCallback } from 'react';
import { Course, Workout, User } from '../types/interfaces';
import { getAllCourses, addCourseToUser, getUserCourses as getUserCoursesAPI, getCourseById, getWorkoutById } from '../api/courses';
import { getUserProfile as getUserProfileFromAPI, updateUserProfile } from '../api/user';
import { useAuth } from './useAuth';
import { database } from '../config/firebase';
import { ref, get, remove } from "firebase/database";

const CACHE_EXPIRATION_TIME = 8 * 60 * 60 * 1000; // 8 часов в миллисекундах
interface CachedData<T> {
  data: T;
  timestamp: number;
}

const getCachedData = <T>(key: string): T | null => {
  try {
    const cachedItem = localStorage.getItem(key);
    if (cachedItem) {
      const { data, timestamp }: CachedData<T> = JSON.parse(cachedItem);
      if (Date.now() - timestamp < CACHE_EXPIRATION_TIME) {
        return data;
      }
    }
  } catch (error) {
    console.error('Error reading from localStorage:', error);
  }
  return null;
};

const setCachedData = <T>(key: string, data: T) => {
  try {
    const cacheItem: CachedData<T> = {
      data,
      timestamp: Date.now(),
    };
    localStorage.setItem(key, JSON.stringify(cacheItem));
  } catch (error) {
    console.error('Error writing to localStorage:', error);
  }
};

export const useCourses = () => {
  const [courses, setCourses] = useState<Course[]>([]);
  const [userCourses, setUserCourses] = useState<Course[]>([]);
  const [loading, setLoading] = useState(true);
  const { user: authUser } = useAuth();

  const fetchCourses = useCallback(async () => {
    setLoading(true);
    try {
      const cachedCourses = getCachedData<Course[]>('allCourses');
      if (cachedCourses) {
        setCourses(cachedCourses);
      } else {
        const allCourses = await getAllCourses();
        setCourses(allCourses);
        setCachedData('allCourses', allCourses);
      }

      if (authUser) {
        const cachedUserCourses = getCachedData<Course[]>(`userCourses_${authUser.uid}`);
        if (cachedUserCourses) {
          setUserCourses(cachedUserCourses);
        } else {
          const userCoursesData = await getUserCoursesAPI(authUser.uid);
          setUserCourses(userCoursesData);
          setCachedData(`userCourses_${authUser.uid}`, userCoursesData);
        }
      } else {
        setUserCourses([]);
      }
    } catch (error) {
      console.error('Error fetching courses:', error);
    } finally {
      setLoading(false);
    }
  }, [authUser]);

  useEffect(() => {
    fetchCourses();
  }, [fetchCourses]);

  const addCourse = async (courseId: string) => {
    if (!authUser) return;
    try {
      await addCourseToUser(authUser.uid, courseId);
      const updatedUserCourses = await getUserCoursesAPI(authUser.uid);
      setUserCourses(updatedUserCourses);
    } catch (error) {
      console.error('Error adding course:', error);
      throw error;
    }
  };

  const getCourse = async (courseId: string): Promise<Course | null> => {
    try {
      return await getCourseById(courseId);
    } catch (error) {
      console.error('Error fetching course:', error);
      return null;
    }
  };

  const getWorkout = useCallback(async (workoutId: string): Promise<Workout | null> => {
    try {
      return await getWorkoutById(workoutId);
    } catch (error) {
      console.error('Error fetching workout:', error);
      return null;
    }
  }, []);

  const getUserProfile = async (): Promise<User | null> => {
    if (!authUser) return null;
    try {
      return await getUserProfileFromAPI(authUser.uid);
    } catch (error) {
      console.error('Error fetching user profile:', error);
      return null;
    }
  };

  const updateProfile = async (data: Partial<User>): Promise<void> => {
    if (!authUser) return;
    try {
      await updateUserProfile(authUser.uid, data);
    } catch (error) {
      console.error('Error updating user profile:', error);
      throw error;
    }
  };
  
  const getUserCourses = useCallback(async (userId: string): Promise<Course[]> => {
    try {
      return await getUserCoursesAPI(userId);
    } catch (error) {
      console.error('Error fetching user courses:', error);
      return [];
    }
  }, []);

  const resetCourseProgress = async (userId: string, courseId: string) => {
    if (!userId || !courseId) return;
    try {
      const courseRef = ref(database, `courses/courses/${courseId}`);
      const courseSnapshot = await get(courseRef);
      if (courseSnapshot.exists()) {
        const courseData = courseSnapshot.val();
        const workoutIds = courseData.workouts || [];
        const progressRef = ref(database, `userProgress/${userId}`);
        const progressSnapshot = await get(progressRef);
        if (progressSnapshot.exists()) {
          const progressData = progressSnapshot.val();
          for (const workoutId of workoutIds) {
            if (progressData[workoutId]) {
              await remove(ref(database, `userProgress/${userId}/${workoutId}`));
            }
          }
        }
      }
    } catch (error) {
      console.error('Error resetting course progress:', error);
    }
  };

  return {
    courses,
    userCourses,
    loading,
    addCourse,
    getCourse,
    getWorkout,
    getUserProfile,
    updateProfile,
    getUserCourses,
    resetCourseProgress,
  };
};