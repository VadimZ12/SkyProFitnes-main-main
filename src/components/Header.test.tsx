import React from 'react'
import { render } from '@testing-library/react'
import Header from './Header'
import { useAuth } from './../hooks/useAuth'
import { MemoryRouter } from 'react-router-dom'
import '@testing-library/jest-dom'

jest.mock('./../hooks/useAuth')
jest.mock('../config/firebase', () => ({
	auth: { signInWithEmailAndPassword: jest.fn(), signOut: jest.fn() },
	database: { ref: jest.fn(), get: jest.fn(), set: jest.fn() },
}))

describe('Header component', () => {
	const mockOnLoginClick = jest.fn()

	beforeAll(() => {
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		;(global as any).importMeta = {
			env: {
				VITE_FIREBASE_API_KEY: 'mock-api-key',
				VITE_FIREBASE_AUTH_DOMAIN: 'mock-auth-domain',
				VITE_FIREBASE_DATABASE_URL: 'mock-database-url',
				VITE_FIREBASE_PROJECT_ID: 'mock-project-id',
				VITE_FIREBASE_STORAGE_BUCKET: 'mock-storage-bucket',
				VITE_FIREBASE_MESSAGING_SENDER_ID: 'mock-messaging-sender-id',
				VITE_FIREBASE_APP_ID: 'mock-app-id',
			},
		}
	})

	it('renders correctly when the user is not logged in', () => {
		;(useAuth as jest.Mock).mockReturnValue({ user: null })

		const { asFragment, getByText } = render(
			<MemoryRouter>
				<React.Fragment>
					<Header onLoginClick={mockOnLoginClick} />
				</React.Fragment>
			</MemoryRouter>
		)

		expect(getByText('Войти')).toBeInTheDocument()
		expect(asFragment()).toMatchSnapshot()
	})

	it('renders correctly when the user is logged in', () => {
		;(useAuth as jest.Mock).mockReturnValue({
			user: {
				displayName: 'John Doe',
				email: 'johndoe@example.com',
				photoURL: '/images/profile.png',
			},
		})

		const { asFragment, getByText, getByAltText } = render(
			<MemoryRouter>
				<React.Fragment>
					<Header onLoginClick={mockOnLoginClick} />
				</React.Fragment>
			</MemoryRouter>
		)

		expect(getByText('John Doe')).toBeInTheDocument()
		expect(getByAltText('user avatar')).toBeInTheDocument()
		expect(asFragment()).toMatchSnapshot()
	})
})
