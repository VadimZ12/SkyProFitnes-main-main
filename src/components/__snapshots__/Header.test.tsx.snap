// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly when the user is logged in 1`] = `
<DocumentFragment>
  <header
    class="flex justify-between mt-[50px] items-center"
  >
    <a
      class="flex flex-col gap-[15px]"
      href="/"
    >
      <img
        alt="logo"
        class="w-[220px] h-[35px]"
        loading="lazy"
        src="test-file-stub"
      />
      <p
        class="font-roboto text-[18px] font-normal leading-[19.8px] text-left text-gray mobile:hidden"
      >
        Онлайн-тренировки для занятий дома
      </p>
    </a>
    <div
      class="relative"
    >
      <div
        class="flex items-center cursor-pointer gap-5"
      >
        <img
          alt="user avatar"
          class="w-[41.67px] h-[41.67px] p-1 rounded-full bg-[#D9D9D9] object-cover"
          src="/images/profile.png"
        />
        <p
          class="font-roboto text-[24px] font-normal leading-[110%] text-black"
        >
          John Doe
        </p>
        <svg
          class="w-[13px] h-[13px]"
          fill="none"
          viewBox="0 0 14 7"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 1L7 6L13 1"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header component renders correctly when the user is not logged in 1`] = `
<DocumentFragment>
  <header
    class="flex justify-between mt-[50px] items-center"
  >
    <a
      class="flex flex-col gap-[15px]"
      href="/"
    >
      <img
        alt="logo"
        class="w-[220px] h-[35px]"
        loading="lazy"
        src="test-file-stub"
      />
      <p
        class="font-roboto text-[18px] font-normal leading-[19.8px] text-left text-gray mobile:hidden"
      >
        Онлайн-тренировки для занятий дома
      </p>
    </a>
    <div
      class="relative"
    >
      <button
        class="font-roboto h-[52px] font-normal text-lg leading-[110%] text-center rounded-[46px] transition-all duration-300 ease-in-out cursor-pointer bg-primary text-black hover:bg-primary-hover active:bg-black active:text-white w-[103px] h-[52px]"
        type="button"
      >
        Войти
      </button>
    </div>
  </header>
</DocumentFragment>
`;
